libsolidity/AST.cpp

```diff
@@ -485,11 +485,24 @@ void FunctionCall::checkTypeRequirements()
 		// and then ask if that is implicitly convertible to the struct represented by the
 		// function parameters
 		TypePointers const& parameterTypes = functionType->getParameterTypes();
-		if (parameterTypes.size() != m_arguments.size())
+		if (functionType->getLocation() !=FunctionType::Location::SHA3 && parameterTypes.size() != m_arguments.size())
 			BOOST_THROW_EXCEPTION(createTypeError("Wrong argument count for function call."));
 		for (size_t i = 0; i < m_arguments.size(); ++i)
-			if (!m_arguments[i]->getType()->isImplicitlyConvertibleTo(*parameterTypes[i]))
+		{
+			if (functionType->getLocation() == FunctionType::Location::SHA3)
+			{
+#if 0 // are we sure we want that? Literal constant nums can't live outside storage and so sha3(42) will fail
```

Comment:
I would rather not convert everything to hash256. For example I think it is expected that `sha3("")` is the sha3 of the empty string and not `sha3(0)`. Everything should be copied to memory and take exactly "calldataEncodedSize" bytes. Note that there is a special case for IntegerConstants.